tools/clang/lib/Sema/SemaDecl.cpp

```diff
@@ -9153,6 +9153,28 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init,
   // Get the decls type and save a reference for later, since
   // CheckInitializerTypes may change it.
   QualType DclT = VDecl->getType(), SavT = DclT;
+
+  // HLSL Change begin
+  // When initializing an HLSL resource type we should diagnose mismatches in
+  // globally coherent annotations _unless_ the source is a dynamic resource
+  // placeholder type where we safely infer the globallycoherent annotaiton.
+  if (getLangOpts().HLSL) {
+    QualType SrcTy = Init->getType();
+    QualType DstTy = DclT;
+    if (SrcTy->isArrayType() && DstTy->isArrayType()) {
```

Comment:
The code here feels similar as CheckGloballyCoherentMismatch, is it possible to modify CheckGloballyCoherentMismatch and share the code?

Reply:
Fixed in 26728cd488ce0c06b7828b10e6e814a6e28ff6bf